Home page, structure and formatting improvements (#54)
nhsbsa-data-analytics/nhsbsa-analytical-code-assurance-playbook: guidance/02-tested/index.qmd

### What is *automated testing*?
In software development workflows, *testing* and *automated testing* are often used interchangeably to mean the latter. 

Automated testing means writing tests in code that run automatically. This approach is repeatable, gives quick feedback on changes, and catches issues early. However, it takes time to write and maintain, and isn't always the best way to test everything.

We recommend combining manual and automated testing, using automated testing where it adds the most value.

## Why should we test our analysis?

Testing is a fundamental part of quality assurance, helping to:

* Verify that the code works as intended.
* Check that the outputs of the analysis are correct and reproducible.
* Creating an audit trail which provides evidence that the analysis produces the correct output.
* Catch errors and bugs before the code is used to produce results.
* [Document][documented] how the code is intended to behave, promoting collaboration and reproducibility.
* Act as a safety net when making changes or improvements.

> * Encourage more [modular][modular] designs which are easier to test and maintain.

Reviewer: ```suggestion
* verify that the code works as intended.
* check that the outputs of the analysis are correct and reproducible.
* creating an audit trail which provides evidence that the analysis produces the correct output.
* catch errors and bugs before the code is used to produce results.
* [document][documented] how the code is intended to behave, promoting collaboration and reproducibility.
* act as a safety net when making changes or improvements.
* encourage more [modular][modular] designs which are easier to test and maintain.
```
removing capitalisation in line with GDS web writing guidelines

## How do we test our analysis?

> We recommend adopting proportionate (as judged by peer-review), risk-based approach with more focus on testing logic that is newer, more complex, or business critical.

Reviewer: 'as judged by peer-review' - I haven't reviewed peer review page yet, but is this referencing specific section there? if so would recommend adding a link. if not, what does this actually entail?

Author: Happy to remove, but the point was that 'proportionate' isn't a strict definition so it's to be agreed with a peer what that means.

Reviewer: maybe just rewording slightly to make more explicit? 
```suggestion
We recommend adopting a proportionate, risk-based approach with more focus on testing logic that is newer, more complex, or business critical. Your approach should be discussed with your team, critical friend, or peer reviewer to agree that it is in-line with the identified risk profile of the project.
```
that suggestion is a bit wordy still but something along those lines?

### Roles and responsibilities

Testing is a shared responsibility, and everyone involved in the analysis should be encouraged to contribute to testing:

* The analyst writing the code should be responsible for writing and running tests to verify their code works as expected.
* [Peer reviewers][peer-review] should check the evidence that appropraite tests have been carried out including thatthe right things have been tested. They should also reproduce the test results as part of their review process.

> * If you have a dedicated testing or quality assurance role, they should be responsible for overseeing the testing process, providing guidance and support to analysts and reviewers, and ensuring that testing is being done effectively across the team.

Reviewer: ```suggestion
* the analyst writing the code should be responsible for writing and running tests to verify their code works as expected.
* [peer reviewers][peer-review] should check the evidence that appropriate tests have been carried out including that the right things have been tested. They should also reproduce the test results as part of their review process.
* if you have a dedicated testing or quality assurance role, they should be responsible for overseeing the testing process, providing guidance and support to analysts and reviewers, and ensuring that testing is being done effectively across the team.
```
as previous bullets and amending a couple typos

### Recommended workflow 
We recommend the following workflow for testing analytical code:

1. Identify the scenarios or behaviours you want to test. For example, this could be checking the result of a specific calculation or data transformation.  
2. Define the inputs and expected outputs or behaviour. Tip: Test *what* the code should do, rather than *how* it does it. 
3. Run the code that you are testing and observe the results.
4. Assert that the code did what you expected it to do (usually comparing the actual output matches the expected output). 
5. (If needed) Fix any errors or issues.

> 6. (If needed) Rerun the tests to verify that your code works as expected.

Reviewer: ```suggestion
1. identify the scenarios or behaviours you want to test. For example, this could be checking the result of a specific calculation or data transformation.  
2. define the inputs and expected outputs or behaviour. Tip: Test *what* the code should do, rather than *how* it does it. 
3. run the code that you are testing and observe the results.
4. assert that the code did what you expected it to do (usually comparing the actual output matches the expected output). 
5. (if needed) fix any errors or issues.
6. (if needed) rerun the tests to verify that your code works as expected.
```
capitalisation

<!-- AJ: Commented out - not sure we need this level of detail when it's available in links.
### Additional details on types of tests

In analytical work, the following types of tests are often the most important to consider:

| Type | Description | Benefits |
| ----- | ------------- | -------- |
| Unit Tests | Small isolated tests to ensure each individual component (e.g. a single calculation) works as expected | Help you locate and identify the root cause of any issues or errors |
| Integration Tests | Testing interactions between two or more individual components (for example, can the data output by one calculation be used as the input to another?) | Verify that different parts of the analysis work together as expected |
| End-to-end Tests | Testing the entire analysis from start to finish | Verify that the whole process produces the expected results, given certain inputs |

In practice, a mix of all three types often works best. 

* For analytical work, integration and end-to-end tests tend to give the most value, as they verify the analysis works correctly without requiring a lot of maintenance as the analysis changes.

> * As unit tests are the most costly to maintain, it is recommended that they target the most complex and critical logic, and that they focus on testing the *behaviour* of the code rather than the *implementation*. -->

Reviewer: agree that detail is available elsewhere, but do think that testing will be the area that people want/need the most guidance on and we should make the content as easily accessible as possible. maybe not for this v1 but next revision i'd like to include detailed examples of implementing these kinds of test in an analytical context.

Author: Agree - think it's V2 to set this out more in practical terms than just saying what these types of tests are.

Author: Added an issue to capture so will resolve #55

### Additional guidance for automated testing
If you are producing analysis using code, the following points can help you go even further.

* **Use Testing Frameworks:** Use established testing frameworks appropriate for the coding language.
* **Automate Test Execution:** Run tests automatically, especially as part of a Continuous Integration (CI) pipeline, to get fast feedback on changes.

> It is important to also ensure that new changes or bug fixes haven't introduced a bug in code that was already tested successfully. This is a good reason to regularly run all existing unit, integration and E2E tests, which is much easier if you have automated tests in place and easier still if you have a CI pipeline set up.

Reviewer: ```suggestion
If you are producing analysis using code, you can take this approach further by using:

* established testing frameworks for you the coding language you're using
* automated test execution to run tests automatically, especially as part of a continuous integration (CI) pipline

 It is important to make sure that new changes or bug fixes haven't introduced new bugs in code that was already tested successfully. This is a good reason to automate as many tests as possible and regularly run them as a suite of tests.
```
this bit had a strange tone to it. have do a little rewrite
